### PR TITLE
NDLC - Perform low resolution scope timing

### PR DESCRIPTION
**Pull request by snakefoot**
Implemented async scope timing

### CODE REVIEW

File: src/NLog/LayoutRenderers/NdlcTimingLayoutRenderer.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                {
+                    TimeSpan duration = scopeBegin != DateTime.MinValue ? DateTime.UtcNow - scopeBegin : TimeSpan.Zero;
+                    if (duration < TimeSpan.Zero)
+                        duration = TimeSpan.Zero;
+
+                    var formatProvider = GetFormatProvider(logEvent, null);
+#if !NET3_5
+                    builder.Append(duration.ToString(Format, formatProvider));
+#else
+                    builder.Append(duration.ToString());
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/NdlcTimingLayoutRenderer.cs]: formatProvider not needed here?

File: src/NLog/NestedDiagnosticsLogicalContext.cs [hunk 2]
```diff
@@ -90,6 +90,35 @@ public static object PopObject()
...
 
         /// <summary>
+        /// Peeks the current scope, and returns its start time
+        /// </summary>
+        /// <returns>Scope Creation Time</returns>
+        public static DateTime PeekCurrentScopeBeginTime()
+        {
+            var currentContext = GetThreadLocal();
+            if (currentContext == null)
+                return DateTime.MinValue;
```
Comment (304NotModified) [inline src/NLog/NestedDiagnosticsLogicalContext.cs]: what about nullable datetimes?

File: src/NLog/NestedDiagnosticsLogicalContext.cs [hunk 3]
```diff
@@ -90,6 +90,35 @@ public static object PopObject()
         }
 
         /// <summary>
+        /// Peeks the current scope, and returns its start time
+        /// </summary>
+        /// <returns>Scope Creation Time</returns>
+        public static DateTime PeekCurrentScopeBeginTime()
```
Comment (304NotModified) [inline src/NLog/NestedDiagnosticsLogicalContext.cs]: could this be internal? (2x)

**Comment by snakefoot** [comment]
> could you please explain me in short why we need this? thx!

If you create a scope to mark that you have started a web-request, then you can use this to extract the timings for this scoped web-request (when logging messages that belongs to the web-request).

I guess one could also implement this by calling NDLC.Push with a complex object, but I guess we then need to provide a PeekLast and PeekFirst to retrieve this object in a custom layout-renderer.

File: src/NLog/LayoutRenderers/NdlcTimingLayoutRenderer.cs (hunk 1 above)
Comment (snakefoot) [reply src/NLog/LayoutRenderers/NdlcTimingLayoutRenderer.cs]: There is no TimeSpan.ToString with Format or formatProvider in NET3_5

File: src/NLog/NestedDiagnosticsLogicalContext.cs (hunk 3 above)
Comment (snakefoot) [reply src/NLog/NestedDiagnosticsLogicalContext.cs]: Fixed

File: src/NLog/NestedDiagnosticsLogicalContext.cs (hunk 2 above)
Comment (snakefoot) [reply src/NLog/NestedDiagnosticsLogicalContext.cs]: I prefer DateTime.MinValue as magic value, don't need another one.

File: src/NLog/LayoutRenderers/NdlcTimingLayoutRenderer.cs (hunk 1 above)
Comment (304NotModified) [reply src/NLog/LayoutRenderers/NdlcTimingLayoutRenderer.cs]: Indeed, but I think the call ` var formatProvider = GetFormatProvider(logEvent, null);` is not needed.

Comment (snakefoot) [reply src/NLog/LayoutRenderers/NdlcTimingLayoutRenderer.cs]: Fixed

**Pull request merged**